crates/sui-core/src/checkpoints/mod.rs

```diff
@@ -2226,36 +2284,70 @@ impl CheckpointService {
             accumulator,
             checkpoint_output,
             notify_aggregator.clone(),
+            cur_built_tx.clone(),
             metrics.clone(),
             max_transactions_per_checkpoint,
             max_checkpoint_size_bytes,
         );
-        tasks.spawn(monitored_future!(builder.run()));
-
-        let aggregator = CheckpointAggregator::new(
-            checkpoint_store.clone(),
-            epoch_store.clone(),
-            notify_aggregator.clone(),
-            certified_checkpoint_output,
-            state.clone(),
-            metrics.clone(),
-        );
-        tasks.spawn(monitored_future!(aggregator.run()));
 
         let last_signature_index = epoch_store
             .get_last_checkpoint_signature_index()
             .expect("should not cross end of epoch");
         let last_signature_index = Mutex::new(last_signature_index);
 
-        let service = Arc::new(Self {
+        Arc::new(Self {
             tables: checkpoint_store,
             notify_builder,
             notify_aggregator,
             last_signature_index,
+            last_built_seq,
+            cur_built_tx,
             metrics,
-        });
+            state: Mutex::new(CheckpointServiceState::Unstarted((builder, aggregator))),
+        })
+    }
+
+    /// Starts the CheckpointService.
+    ///
+    /// This function blocks until the CheckpointBuilder re-builds all checkpoints that had
+    /// been built before the most recent restart. You can think of this as a WAL replay
+    /// operation. Upon startup, we may have a number of consensus commits and resulting
+    /// checkpoints that were built but not committed to disk. We want to reprocess the
+    /// commits and rebuild the checkpoints before starting normal operation.
+    pub async fn spawn(&self) -> JoinSet<()> {
+        let mut tasks = JoinSet::new();
+
+        let (builder, aggregator) = self.state.lock().take();
+        tasks.spawn(monitored_future!(builder.run()));
+        tasks.spawn(monitored_future!(aggregator.run()));
 
-        (service, tasks)
+        loop {
+            if tokio::time::timeout(Duration::from_secs(10), self.wait_for_rebuilt_checkpoints())
+                .await
+                .is_ok()
+            {
+                break;
+            } else {
+                debug_fatal!("Still waiting for checkpoints to be rebuilt");
```

Comment:
is it helpful to crash here even if just in debug mode? no invariant has been violated, it seems like it should be a `warn!` at most?

Reply:
yes it is - while testing DQ this catches various cases where the rebuilding process got stuck.